lib/oneAPIKernels/src/oneAPIKernels.jl

```diff
@@ -0,0 +1,283 @@
+module oneAPIKernels
+
+import oneAPI
+import oneAPI: oneL0
+import StaticArrays
+import KernelAbstractions
+
+export oneAPIDevice
+
+KernelAbstractions.get_device(::oneAPI.oneArray) = oneAPIDevice()
+
+
+const FREE_QUEUES = oneL0.ZeCommandQueue[]
+const QUEUES = oneL0.ZeCommandQueue[]
+const QUEUE_GC_THRESHOLD = Ref{Int}(16)
+
+# This code is loaded after an `@init` step
+if haskey(ENV, "KERNELABSTRACTIONS_QUEUES_GC_THRESHOLD")
+    global QUEUE_GC_THRESHOLD[] = parse(Int, ENV["KERNELABSTRACTIONS_QUEUES_GC_THRESHOLD"])
+end
+
+## Queue GC
+# Simplistic queue gc design in which when we have a total number
+# of queues bigger than a threshold, we start scanning the queues
+# and add them back to the freelist if all work on them has completed.
+# Alternative designs:
+# - Enqueue a host function on the stream that adds the stream back to the freelist
+# - Attach a finalizer to events that adds the stream back to the freelist
+# Possible improvements
+# - Add a background task that occasionally scans all queues
+# - Add a hysterisis by checking a "since last scanned" timestamp
+# - Add locking
+const QUEUE_GC_LOCK = Threads.ReentrantLock()
+function next_queue()
+    ctx = oneAPI.context()
+    dev = oneAPI.device()
+    lock(QUEUE_GC_LOCK) do
+        if !isempty(FREE_QUEUES)
+            return pop!(FREE_QUEUES)
+        end
+
+        if length(QUEUES) > QUEUE_GC_THRESHOLD[]
+            for queue in QUEUES
+                # XXX: can we check if the queue is idle?
+                 push!(FREE_QUEUES, queue)
+            end
+        end
+
+        if !isempty(FREE_QUEUES)
+            return pop!(FREE_QUEUES)
+        end
+
+        queue = oneL0.ZeCommandQueue(ctx, dev)
+        push!(QUEUES, queue)
+        return queue
+    end
+end
+
+import KernelAbstractions: Event, CPUEvent, NoneEvent, MultiEvent, CPU, GPU, isdone, failed
+
+struct oneAPIDevice <: GPU end
+
+struct oneAPIEvent <: Event
+    event::oneL0.ZeEvent
+end
+
+failed(::oneAPIEvent) = false
+isdone(ev::oneAPIEvent) = Base.isdone(ev.event)
+
+function Event(::oneAPIDevice)
+    ctx = oneAPI.context()
+    dev = oneAPI.device()
+    event_pool = oneAPI.ZeEventPool(ctx, 1, dev)
+    # XXX: this is inefficient
+    event = event_pool[1]
+    # XXX: other back-ends record/signal here straight away. why?
```

Comment:
The goal here is to have a Event that can be used to synchronize against the default stream.

E.g. if I used GPUArrays I will need to create an event to depend on, to make sure that KA runs after those kernelrs

Reply:
Ah, so it needs to be a globally-synchronized event for interop with other task-local packages.